File: antd-cpp/tests/test_client.cpp
#define DOCTEST_CONFIG_IMPLEMENT_WITH_MAIN
#include <doctest/doctest.h>

#include <nlohmann/json.hpp>

#include "antd/errors.hpp"
#include "antd/models.hpp"
#include "base64.hpp"

using json = nlohmann::json;

// ---------------------------------------------------------------------------
// Base64 encode / decode
// ---------------------------------------------------------------------------

TEST_CASE("base64 encode empty") {
    std::vector<uint8_t> data;
    CHECK(antd::detail::base64_encode(data).empty());
}

TEST_CASE("base64 round-trip") {
    std::string original = "Hello, Autonomi!";
    std::vector<uint8_t> data(original.begin(), original.end());

    auto encoded = antd::detail::base64_encode(data);
    CHECK(encoded == "SGVsbG8sIEF1dG9ub21pIQ==");

    auto decoded = antd::detail::base64_decode(encoded);
    std::string result(decoded.begin(), decoded.end());
    CHECK(result == original);
}

TEST_CASE("base64 round-trip various lengths") {
    // Test padding cases: length % 3 == 0, 1, 2
    for (size_t len : {0, 1, 2, 3, 4, 5, 6, 100, 255}) {
        std::vector<uint8_t> data(len);
        for (size_t i = 0; i < len; ++i) {
            data[i] = static_cast<uint8_t>(i & 0xFF);
        }
        auto encoded = antd::detail::base64_encode(data);
        auto decoded = antd::detail::base64_decode(encoded);
        CHECK(decoded == data);
    }
}

// ---------------------------------------------------------------------------
// Error mapping
// ---------------------------------------------------------------------------

TEST_CASE("error_for_status throws correct types") {
    CHECK_THROWS_AS(antd::error_for_status(400, "bad"), antd::BadRequestError);
    CHECK_THROWS_AS(antd::error_for_status(402, "pay"), antd::PaymentError);
    CHECK_THROWS_AS(antd::error_for_status(404, "missing"), antd::NotFoundError);
    CHECK_THROWS_AS(antd::error_for_status(409, "exists"), antd::AlreadyExistsError);
    CHECK_THROWS_AS(antd::error_for_status(413, "big"), antd::TooLargeError);
    CHECK_THROWS_AS(antd::error_for_status(500, "oops"), antd::InternalError);
    CHECK_THROWS_AS(antd::error_for_status(502, "net"), antd::NetworkError);
    CHECK_THROWS_AS(antd::error_for_status(503, "unknown"), antd::AntdError);
}

TEST_CASE("error status_code is preserved") {
    try {
        antd::error_for_status(404, "not found");
        FAIL("should have thrown");
    } catch (const antd::NotFoundError& e) {
        CHECK(e.status_code == 404);
        CHECK(std::string(e.what()).find("not found") != std::string::npos);
    }
}

TEST_CASE("AntdError subtypes are catchable as AntdError") {
    try {
        antd::error_for_status(500, "boom");
        FAIL("should have thrown");
    } catch (const antd::AntdError& e) {
        CHECK(e.status_code == 500);
    }
}

// ---------------------------------------------------------------------------
// Model JSON parsing (manual, mirrors what client.cpp does)
// ---------------------------------------------------------------------------

TEST_CASE("HealthStatus from JSON") {
    auto j = json::parse(R"({
        "status":"ok",
        "network":"local",
        "version":"0.4.0",
        "evm_network":"local",
        "uptime_seconds":42,
        "build_commit":"abcdef123456",
        "payment_token_address":"0xtoken",
        "payment_vault_address":"0xvault"
    })");
    antd::HealthStatus h{
        .ok = j.value("status", "") == "ok",
        .network = j.value("network", ""),
        .version = j.value("version", ""),
        .evm_network = j.value("evm_network", ""),
        .uptime_seconds = j.value<std::uint64_t>("uptime_seconds", 0),
        .build_commit = j.value("build_commit", ""),
        .payment_token_address = j.value("payment_token_address", ""),
        .payment_vault_address = j.value("payment_vault_address", ""),
    };

    CHECK(h.ok);
    CHECK(h.network == "local");
    CHECK(h.version == "0.4.0");
    CHECK(h.evm_network == "local");
    CHECK(h.uptime_seconds == 42);
    CHECK(h.build_commit == "abcdef123456");
    CHECK(h.payment_token_address == "0xtoken");
    CHECK(h.payment_vault_address == "0xvault");
}

TEST_CASE("HealthStatus defaults populate when pre-0.4.0 daemon omits diagnostics") {
    auto j = json::parse(R"({"status":"ok","network":"default"})");
    antd::HealthStatus h{
        .ok = j.value("status", "") == "ok",
        .network = j.value("network", ""),
        .version = j.value("version", ""),
        .evm_network = j.value("evm_network", ""),
        .uptime_seconds = j.value<std::uint64_t>("uptime_seconds", 0),
        .build_commit = j.value("build_commit", ""),
        .payment_token_address = j.value("payment_token_address", ""),
        .payment_vault_address = j.value("payment_vault_address", ""),
    };

    CHECK(h.ok);
    CHECK(h.network == "default");
    CHECK(h.version.empty());
    CHECK(h.evm_network.empty());
    CHECK(h.uptime_seconds == 0);
    CHECK(h.build_commit.empty());
}

TEST_CASE("PutResult from JSON") {
    auto j = json::parse(R"({"cost":"100","address":"abc123"})");
    antd::PutResult r;
    r.cost = j.value("cost", "");
    r.address = j.value("address", "");

    CHECK(r.cost == "100");
    CHECK(r.address == "abc123");
}

// ---------------------------------------------------------------------------
// PrepareUploadResult: merkle payment parsing
// ---------------------------------------------------------------------------

TEST_CASE("PrepareUploadResult merkle from JSON") {
    auto j = json::parse(R"({
        "upload_id": "mup1",
        "payment_type": "merkle",
        "depth": 5,
        "merkle_payment_timestamp": 1712150400,
        "total_amount": "0",
        "payment_vault_address": "0xvault",
        "payment_token_address": "0xtoken",
        "rpc_url": "http://rpc.example.com",
        "pool_commitments": [
            {
                "pool_hash": "0xaabbccdd",
                "candidates": [
                    {"rewards_address": "0x1111", "amount": "500"},
                    {"rewards_address": "0x2222", "amount": "600"}
                ]
            }
        ]
    })");

    // Simulate what parse_prepare_response does (mirrors client.cpp logic)
    antd::PrepareUploadResult r;
    r.upload_id = j.value("upload_id", "");
    r.payment_type = j.value("payment_type", "");
    r.total_amount = j.value("total_amount", "");
    r.payment_vault_address = j.value("payment_vault_address", "");
    r.payment_token_address = j.value("payment_token_address", "");
    r.rpc_url = j.value("rpc_url", "");

    if (r.payment_type.empty()) {
        r.payment_type = "wave_batch";
    }

    if (j.contains("payments") && j["payments"].is_array()) {
        for (const auto& p : j["payments"]) {
            if (p.is_object()) {
                r.payments.push_back(antd::PaymentInfo{
                    p.value("quote_hash", ""),
                    p.value("rewards_address", ""),
                    p.value("amount", ""),
                });
            }
        }
    }

    if (r.payment_type == "merkle") {
        r.depth = j.value("depth", 0);
        r.merkle_payment_timestamp = j.value("merkle_payment_timestamp", uint64_t{0});

        if (j.contains("pool_commitments") && j["pool_commitments"].is_array()) {
            for (const auto& pc : j["pool_commitments"]) {
                if (!pc.is_object()) continue;
                antd::PoolCommitmentEntry entry;
                entry.pool_hash = pc.value("pool_hash", "");
                if (pc.contains("candidates") && pc["candidates"].is_array()) {
                    for (const auto& c : pc["candidates"]) {
                        if (c.is_object()) {
                            entry.candidates.push_back(antd::CandidateNodeEntry{
                                c.value("rewards_address", ""),
                                c.value("amount", ""),
                            });
                        }
                    }
                }
                r.pool_commitments.push_back(std::move(entry));
            }
        }
    }

    CHECK(r.upload_id == "mup1");
    CHECK(r.payment_type == "merkle");
    CHECK(r.depth == 5);
    CHECK(r.merkle_payment_timestamp == 1712150400);
    CHECK(r.total_amount == "0");
    CHECK(r.payment_vault_address == "0xvault");
    CHECK(r.payment_token_address == "0xtoken");
    CHECK(r.rpc_url == "http://rpc.example.com");
    CHECK(r.payments.empty());

    REQUIRE(r.pool_commitments.size() == 1);
    CHECK(r.pool_commitments[0].pool_hash == "0xaabbccdd");
    REQUIRE(r.pool_commitments[0].candidates.size() == 2);
    CHECK(r.pool_commitments[0].candidates[0].rewards_address == "0x1111");
    CHECK(r.pool_commitments[0].candidates[0].amount == "500");
    CHECK(r.pool_commitments[0].candidates[1].rewards_address == "0x2222");
    CHECK(r.pool_commitments[0].candidates[1].amount == "600");
}

TEST_CASE("PrepareUploadResult merkle finalize JSON") {
    // Verify finalize_merkle_upload request body structure
    json body = {
        {"upload_id", "mup1"},
        {"winner_pool_hash", "0xwinnerhash"},
        {"store_data_map", true},
    };

    CHECK(body["upload_id"] == "mup1");
    CHECK(body["winner_pool_hash"] == "0xwinnerhash");
    CHECK(body["store_data_map"] == true);

    // Verify response parsing
    auto resp = json::parse(R"({
        "data_map": "dm_merkle",
        "address": "0xaddr",
        "chunks_stored": 100
    })");

    antd::FinalizeUploadResult fr;
    fr.data_map = resp.value("data_map", "");
    fr.address = resp.value("address", "");
    fr.chunks_stored = resp.value("chunks_stored", int64_t{0});

    CHECK(fr.data_map == "dm_merkle");
    CHECK(fr.address == "0xaddr");
    CHECK(fr.chunks_stored == 100);
}

TEST_CASE("PrepareUploadResult backward compat - no payment_type defaults to wave_batch") {
    // Older daemons do not return payment_type; should default to wave_batch
    auto j = json::parse(R"({
        "upload_id": "up_old",
        "total_amount": "5000",
        "payment_vault_address": "0xvault",
        "payment_token_address": "0xtoken",
        "rpc_url": "http://rpc.old.com",
        "payments": [
            {"quote_hash": "qh1", "rewards_address": "0xr1", "amount": "2500"},
            {"quote_hash": "qh2", "rewards_address": "0xr2", "amount": "2500"}
        ]
    })");

    antd::PrepareUploadResult r;
    r.upload_id = j.value("upload_id", "");
    r.payment_type = j.value("payment_type", "");
    r.total_amount = j.value("total_amount", "");
    r.payment_vault_address = j.value("payment_vault_address", "");
    r.payment_token_address = j.value("payment_token_address", "");
    r.rpc_url = j.value("rpc_url", "");

    if (r.payment_type.empty()) {
        r.payment_type = "wave_batch";
    }

    if (j.contains("payments") && j["payments"].is_array()) {
        for (const auto& p : j["payments"]) {
            if (p.is_object()) {
                r.payments.push_back(antd::PaymentInfo{
                    p.value("quote_hash", ""),
                    p.value("rewards_address", ""),
                    p.value("amount", ""),
                });
            }
        }
    }

    CHECK(r.payment_type == "wave_batch");
    CHECK(r.upload_id == "up_old");
    CHECK(r.total_amount == "5000");
    CHECK(r.payment_vault_address == "0xvault");
    REQUIRE(r.payments.size() == 2);
    CHECK(r.payments[0].quote_hash == "qh1");
    CHECK(r.payments[1].quote_hash == "qh2");

    // Merkle fields should be at defaults
    CHECK(r.depth == 0);
    CHECK(r.merkle_payment_timestamp == 0);
    CHECK(r.pool_commitments.empty());
}

TEST_CASE("FinalizeUploadResult data_map field") {
    // Verify the data_map field is present on FinalizeUploadResult
    antd::FinalizeUploadResult fr;
    fr.data_map = "0xdeadbeef";
    fr.address = "";
    fr.chunks_stored = 42;

    CHECK(fr.data_map == "0xdeadbeef");
    CHECK(fr.address.empty());
    CHECK(fr.chunks_stored == 42);
}

// ---------------------------------------------------------------------------
// NOTE: Full integration tests require a running antd daemon.
// The tests above validate JSON parsing, base64, and error mapping
// without network access.
// ---------------------------------------------------------------------------

File: antd-ruby/test/test_client.rb
# frozen_string_literal: true

require_relative "test_helper"
require "base64"

class TestClient < Minitest::Test
  BASE = "http://localhost:8082"

  def setup
    @client = Antd::Client.new(base_url: BASE)
  end

  # --- Health ---

  def test_health
    body = {
      status: "ok", network: "local",
      version: "0.4.0", evm_network: "local", uptime_seconds: 42,
      build_commit: "abcdef123456",
      payment_token_address: "0xtoken", payment_vault_address: "0xvault"
    }.to_json
    stub_request(:get, "#{BASE}/health")
      .to_return(status: 200, body: body,
                 headers: { "Content-Type" => "application/json" })

    h = @client.health
    assert h.ok
    assert_equal "local", h.network
    assert_equal "0.4.0", h.version
    assert_equal "local", h.evm_network
    assert_equal 42, h.uptime_seconds
    assert_equal "abcdef123456", h.build_commit
    assert_equal "0xtoken", h.payment_token_address
    assert_equal "0xvault", h.payment_vault_address
  end

  # Pre-0.4.0 daemons reply with just status + network — verify the
  # diagnostic fields default to empty / 0 rather than NPE-ing.
  def test_health_pre_0_4_0_daemon
    stub_request(:get, "#{BASE}/health")
      .to_return(status: 200, body: '{"status":"ok","network":"default"}',
                 headers: { "Content-Type" => "application/json" })

    h = @client.health
    assert h.ok
    assert_equal "default", h.network
    assert_equal "", h.version
    assert_equal "", h.evm_network
    assert_equal 0, h.uptime_seconds
  end

  # --- Data Public ---

  def test_data_put_public
    stub_request(:post, "#{BASE}/v1/data/public")
      .to_return(status: 200, body: '{"cost":"100","address":"abc123"}',
                 headers: { "Content-Type" => "application/json" })

    result = @client.data_put_public("hello")
    assert_equal "100", result.cost
    assert_equal "abc123", result.address
  end

  def test_data_get_public
    encoded = Base64.strict_encode64("hello")
    stub_request(:get, "#{BASE}/v1/data/public/abc123")
      .to_return(status: 200, body: %({"data":"#{encoded}"}),
                 headers: { "Content-Type" => "application/json" })

    data = @client.data_get_public("abc123")
    assert_equal "hello", data
  end

  # --- Data Private ---

  def test_data_put_private
    stub_request(:post, "#{BASE}/v1/data/private")
      .to_return(status: 200, body: '{"cost":"200","data_map":"dm123"}',
                 headers: { "Content-Type" => "application/json" })

    result = @client.data_put_private("secret")
    assert_equal "200", result.cost
    assert_equal "dm123", result.address
  end

  def test_data_get_private
    encoded = Base64.strict_encode64("secret")
    stub_request(:get, %r{#{BASE}/v1/data/private\?data_map=})
      .to_return(status: 200, body: %({"data":"#{encoded}"}),
                 headers: { "Content-Type" => "application/json" })

    data = @client.data_get_private("dm123")
    assert_equal "secret", data
  end

  # --- Data Cost ---

  def test_data_cost
    stub_request(:post, "#{BASE}/v1/data/cost")
      .to_return(status: 200,
                 body: '{"cost":"50","file_size":4,"chunk_count":3,"estimated_gas_cost_wei":"150000000000000","payment_mode":"single"}',
                 headers: { "Content-Type" => "application/json" })

    est = @client.data_cost("test")
    assert_equal "50", est.cost
    assert_equal 4, est.file_size
    assert_equal 3, est.chunk_count
    assert_equal "150000000000000", est.estimated_gas_cost_wei
    assert_equal "single", est.payment_mode
  end

  # --- Chunks ---

  def test_chunk_put
    stub_request(:post, "#{BASE}/v1/chunks")
      .to_return(status: 200, body: '{"cost":"10","address":"chunk1"}',
                 headers: { "Content-Type" => "application/json" })

    result = @client.chunk_put("chunkdata")
    assert_equal "10", result.cost
    assert_equal "chunk1", result.address
  end

  def test_chunk_get
    encoded = Base64.strict_encode64("chunkdata")
    stub_request(:get, "#{BASE}/v1/chunks/chunk1")
      .to_return(status: 200, body: %({"data":"#{encoded}"}),
                 headers: { "Content-Type" => "application/json" })

    data = @client.chunk_get("chunk1")
    assert_equal "chunkdata", data
  end

  # --- Files ---

  def test_file_upload_public
    stub_request(:post, "#{BASE}/v1/files/upload/public")
      .to_return(status: 200,
                 body: '{"address":"file1","storage_cost_atto":"1000","gas_cost_wei":"42","chunks_stored":3,"payment_mode_used":"auto"}',
                 headers: { "Content-Type" => "application/json" })

    result = @client.file_upload_public("/tmp/test.txt")
    assert_equal "file1", result.address
    assert_equal "1000", result.storage_cost_atto
    assert_equal "42", result.gas_cost_wei
    assert_equal 3, result.chunks_stored
    assert_equal "auto", result.payment_mode_used
  end

  def test_file_download_public
    stub_request(:post, "#{BASE}/v1/files/download/public")
      .to_return(status: 200, body: "",
                 headers: { "Content-Type" => "application/json" })

    assert_nil @client.file_download_public("file1", "/tmp/out.txt")
  end

  def test_dir_upload_public
    stub_request(:post, "#{BASE}/v1/dirs/upload/public")
      .to_return(status: 200,
                 body: '{"address":"dir1","storage_cost_atto":"2000","gas_cost_wei":"100","chunks_stored":5,"payment_mode_used":"merkle"}',
                 headers: { "Content-Type" => "application/json" })

    result = @client.dir_upload_public("/tmp/mydir")
    assert_equal "dir1", result.address
    assert_equal "2000", result.storage_cost_atto
    assert_equal "100", result.gas_cost_wei
    assert_equal 5, result.chunks_stored
    assert_equal "merkle", result.payment_mode_used
  end

  def test_dir_download_public
    stub_request(:post, "#{BASE}/v1/dirs/download/public")
      .to_return(status: 200, body: "",
                 headers: { "Content-Type" => "application/json" })

    assert_nil @client.dir_download_public("dir1", "/tmp/outdir")
  end

  def test_file_cost
    stub_request(:post, "#{BASE}/v1/files/cost")
      .to_return(status: 200,
                 body: '{"cost":"1000","file_size":4096,"chunk_count":3,"estimated_gas_cost_wei":"150000000000000","payment_mode":"auto"}',
                 headers: { "Content-Type" => "application/json" })

    est = @client.file_cost("/tmp/test.txt", true)
    assert_equal "1000", est.cost
    assert_equal 4096, est.file_size
    assert_equal 3, est.chunk_count
    assert_equal "150000000000000", est.estimated_gas_cost_wei
    assert_equal "auto", est.payment_mode
  end

  # --- Merkle Batch Payment ---

  def test_prepare_upload_merkle
    response_body = {
      upload_id: "up_merkle_1",
      payment_type: "merkle_batch",
      depth: 3,
      total_amount: "5000",
      payments: [],
      payment_vault_address: "0xMERKLE",
      payment_token_address: "0xTOKEN",
      rpc_url: "http://localhost:8545",
      merkle_payment_timestamp: 1700000000,
      pool_commitments: [
        {
          pool_hash: "pool_abc",
          candidates: [
            { rewards_address: "0xR1", amount: "2000" },
            { rewards_address: "0xR2", amount: "3000" }
          ]
        }
      ]
    }.to_json

    stub_request(:post, "#{BASE}/v1/upload/prepare")
      .to_return(status: 200, body: response_body,
                 headers: { "Content-Type" => "application/json" })

    result = @client.prepare_upload("/tmp/merkle/file.dat")
    assert_instance_of Antd::PrepareUploadResult, result
    assert_equal "up_merkle_1", result.upload_id
    assert_equal "merkle_batch", result.payment_type
    assert_equal 3, result.depth
    assert_equal "5000", result.total_amount
    assert_equal 1700000000, result.merkle_payment_timestamp
    assert_equal "0xMERKLE", result.payment_vault_address
    assert_equal [], result.payments

    assert_equal 1, result.pool_commitments.length
    pc = result.pool_commitments[0]
    assert_instance_of Antd::PoolCommitmentEntry, pc
    assert_equal "pool_abc", pc.pool_hash
    assert_equal 2, pc.candidates.length
    assert_instance_of Antd::CandidateNodeEntry, pc.candidates[0]
    assert_equal "0xR1", pc.candidates[0].rewards_address
    assert_equal "2000", pc.candidates[0].amount
    assert_equal "0xR2", pc.candidates[1].rewards_address
    assert_equal "3000", pc.candidates[1].amount
  end

  def test_finalize_merkle_upload
    stub_request(:post, "#{BASE}/v1/upload/finalize")
      .to_return(status: 200, body: '{"address":"0xFINAL","chunks_stored":42}',
                 headers: { "Content-Type" => "application/json" })

    result = @client.finalize_merkle_upload("up_merkle_1", "pool_abc", store_data_map: true)
    assert_instance_of Antd::FinalizeUploadResult, result
    assert_equal "0xFINAL", result.address
    assert_equal 42, result.chunks_stored
  end

  def test_prepare_upload_backward_compat
    response_body = {
      upload_id: "up_compat_1",
      payments: [
        { quote_hash: "qh1", rewards_address: "0xR1", amount: "100" }
      ],
      total_amount: "100",
      payment_vault_address: "0xDATA",
      payment_token_address: "0xTOKEN",
      rpc_url: "http://localhost:8545"
    }.to_json

    stub_request(:post, "#{BASE}/v1/upload/prepare")
      .to_return(status: 200, body: response_body,
                 headers: { "Content-Type" => "application/json" })

    result = @client.prepare_upload("/tmp/compat/file.dat")
    assert_instance_of Antd::PrepareUploadResult, result
    assert_equal "up_compat_1", result.upload_id
    assert_equal "wave_batch", result.payment_type
    assert_equal 0, result.depth
    assert_equal [], result.pool_commitments
    assert_equal 0, result.merkle_payment_timestamp
    assert_equal "0xDATA", result.payment_vault_address

    assert_equal 1, result.payments.length
    assert_equal "qh1", result.payments[0].quote_hash
  end

  # --- Error Mapping ---

  def test_error_404
    stub_request(:get, "#{BASE}/health")
      .to_return(status: 404, body: '{"error":"not found"}',
                 headers: { "Content-Type" => "application/json" })

    err = assert_raises(Antd::NotFoundError) { @client.health }
    assert_equal 404, err.status_code
    assert_includes err.message, "not found"
  end

  def test_error_400
    stub_request(:get, "#{BASE}/health")
      .to_return(status: 400, body: '{"error":"bad request"}',
                 headers: { "Content-Type" => "application/json" })

    err = assert_raises(Antd::BadRequestError) { @client.health }
    assert_equal 400, err.status_code
  end

  def test_error_402
    stub_request(:get, "#{BASE}/health")
      .to_return(status: 402, body: '{"error":"payment required"}',
                 headers: { "Content-Type" => "application/json" })

    err = assert_raises(Antd::PaymentError) { @client.health }
    assert_equal 402, err.status_code
  end

  def test_error_409
    stub_request(:get, "#{BASE}/health")
      .to_return(status: 409, body: '{"error":"already exists"}',
                 headers: { "Content-Type" => "application/json" })

    err = assert_raises(Antd::AlreadyExistsError) { @client.health }
    assert_equal 409, err.status_code
  end

  def test_error_413
    stub_request(:get, "#{BASE}/health")
      .to_return(status: 413, body: '{"error":"too large"}',
                 headers: { "Content-Type" => "application/json" })

    err = assert_raises(Antd::TooLargeError) { @client.health }
    assert_equal 413, err.status_code
  end

  def test_error_500
    stub_request(:get, "#{BASE}/health")
      .to_return(status: 500, body: '{"error":"internal error"}',
                 headers: { "Content-Type" => "application/json" })

    err = assert_raises(Antd::InternalError) { @client.health }
    assert_equal 500, err.status_code
  end

  def test_error_502
    stub_request(:get, "#{BASE}/health")
      .to_return(status: 502, body: '{"error":"network error"}',
                 headers: { "Content-Type" => "application/json" })

    err = assert_raises(Antd::NetworkError) { @client.health }
    assert_equal 502, err.status_code
  end

  def test_error_unknown_status
    stub_request(:get, "#{BASE}/health")
      .to_return(status: 503, body: '{"error":"unavailable"}',
                 headers: { "Content-Type" => "application/json" })

    err = assert_raises(Antd::AntdError) { @client.health }
    assert_equal 503, err.status_code
  end
end
